### ISSUE TITLE
Discuss type verbosity.

### ISSUE THREAD

**Issue by sledorze** [gcanti/io-ts#165]
I'm experiencing growing pain of consequences of hiding details of runtime types for two reasons: they're too verbose to read and make errors reporting unreadable.

The growing pain come from loosing the runtime structure behind simpler interfaces, and when this structure is needed (generation of api doc, generation tools and others) then the verbosity comes back.

The current situation is very frustrating.

I am now contemplating this declaration:

```typescript
export class InterfaceType<P extends AnyProps, A = any, O = A, I = mixed> extends Type<A, O, I> {
```

and wondering if it should not have better been:

```typescript
export class InterfaceType<P extends AnyProps, I = mixed> extends Type<TypeOfProps<P>, OutputOfProps<P>, I> {
```
This reduce the structure duplication in the type params from 3 to 1 and I'm not thinking of usages requiring those explicit type params (A, 0, I), I feel that it is unnecessary exposure of type parameters that lead to complexity, verbosity and may also introduce potential errors from bad usages.

I may be wrong on the usage part; hence this Issue to understand those and eventually come to a solution if any.

---

Note that we also need to give meaningful Names by using interfaces (like 'User') in place of structural information.
This appears to contradicts the above direction, however, I see that as a separate issue but maybe that's not the case.

**Comment by gcanti** [comment]
> The current situation is very frustrating

I feel your pain, though the patterns suggested here https://github.com/gcanti/io-ts/issues/149#issuecomment-384956527 look promising

> Note that we also need to give meaningful Names by using interfaces (like 'User') in place of structural information. This appears to contradicts the above direction

Indeed, Part I and Part II of this issue look contradictory

**Comment by sledorze** [comment]
Maybe we try to do too much with a single representation; actually both Encoders/Decoders and also Type representation.
The reason we've done it is it be able to not write/maintain both aspects in parallel with minimum work short term, the downside being that it conflates all requirements and becomes too complex long term.

I have many different thinking at the moment:

I use the pattern of giving the same name to the Runtimetype (e.g. `User` encoder/decoder at value level) as to the actual Type (`User` at type level).
This is useful because you can import both the value and type with only one import.
However I wonder if the pattern should not be a `companion` value which purpose would be to aggregate all type related abstractions; io-ts runtime types being one of those and another one being the type representation (one may event think of aggregating specialized type classe instances).

I'm not sure the io-ts DSL of runtimetype specification should be the way to describe types, as in fact it is more about describing the conversions.

I know that those thinking may feel a bit out of topic and a broader subject than io-ts design but they're all related somehow as I think adopting io-ts will lead other developpers into the same situation and improving the experience here, giving guidance of patterns that work (or not) is key.

**Comment by gunzip** [comment]
Don't know if is related, but we're experiencing some issues generating declaration files from 
auto-generated io-ts types (from swagger specs). For example the declarations for this type
[risposta_ppt.ts](https://github.com/teamdigitale/italia-pagopa-api/blob/example-generation/src/types/yaml-to-ts/risposta_ppt.ts) become this one: [risposta_ppt.d.ts](https://github.com/teamdigitale/italia-pagopa-api/blob/example-generation/dist/decls/types/yaml-to-ts/risposta_ppt.d.ts) and it's not the bigger one. As we have a lot of these types typescript goes out of memory after about 30 minutes of compilation time.

**Comment by gcanti** [comment]
@gunzip The proposed solution when generating lots of runtime types is the following pattern

```ts
const _MyType = t.interface({
  faultCode: t.string,
  faultString: t.string,
  id: t.string
})

export interface MyType extends t.TypeOf<typeof _MyType> {}
interface MyTypeOutput extends t.OutputOf<typeof _MyType> {}
interface MyTypeProps extends t.PropsOf<typeof _MyType> {}
export const MyType = t.alias(_MyType)<MyType, MyTypeOutput, MyTypeProps>()
```

**Comment by sgronblo** [comment]
I don't really understand the benefit of the `clean` function. Currently the README mentions that if you hover on nested interfaces you get a verbose type instead of a clean type and then the clean and alias functions are presented as solutions. However the clean type seems to drop all the information about the type and the alias type seems more verbose. There doesn't seem to be any way to get a clean nested type where all the t. prefixed types have been removed?

**Comment by gcanti** [comment]
@sgronblo I'm working on a different solution, see https://github.com/gcanti/io-ts/issues/245#issuecomment-452755553

**Comment by sgronblo** [comment]
Looks promising. Your latest example doesn't show how well it works with the dreaded nested type though :)

**Comment by gcanti** [comment]
You can see the results in the linked declaration file test. The following provides an example

```ts
export const T5 = t.type({
  e: t.type({
    d: t.type({
      c: t.type({
        b: t.type({
          a: t.string
        })
      })
    })
  })
})
/*
const T5: t.TypeC<{
    e: t.TypeC<{
        d: t.TypeC<{
            c: t.TypeC<{
                b: t.TypeC<{
                    a: t.StringC;
                }>;
            }>;
        }>;
    }>;
}>
*/
```

Actually we can apply this technique (intermediary interfaces) with 1.5.x too, I'll release a patch asap.

**Comment by gcanti** [comment]
`t.type` is just an alias of `t.interface` (since `io-ts@1.0.0`)

**Comment by gcanti** [comment]
@sledorze thanks for the feedback

> I really need this line to be doubled to expose the signature accepting Arrays of Types as parameter

What's the issue you are facing? Could you please provide a repro? We might add it to the type-level tests

**Comment by sledorze** [comment]
@gcanti I'm working on a finally tagless encoding of 'Models' which helps creating io-ts, setoids, generators, show instances, etc..

The algebra definition expose the ability to create the intersection from an array of types, that's the reason.

**Comment by gcanti** [comment]
@sledorze let me clarify because I'm a bit puzzled: is the following

> The way intersection is typed is problematic, I really need this line to be doubled to expose the signature accepting Arrays of Types as parameter

a **bug report** about https://github.com/gcanti/io-ts/pull/248 (as I thought it was) or a **feature request**?

**Comment by sledorze** [comment]
@gcanti Indeed; my algebra required the version with an Array.

However the compiler (3.2.2) is (wrongly) not complaining with `io-ts` version 1.5.0 and 1.5.2 (but it should have), not sure why..

Removing the need for an Array does not solve the problem with 1.5.3, so I guess that was a wrong diagnostic and I need more investigations.

**Comment by sledorze** [comment]
Ok just found out I was expecting an overload with one Element while that does not make any sense indeed.
Compiles now :)
Oh.. and indeed it may have been related to #248 actually

**Comment by sgronblo** [comment]
I tried updating io-ts from 1.3.1 to 1.6.2 and rebuild our project. One of our .d.ts files went from a mindblowing 1.5MB to 9kB. Is this to be expected or could I have made some mistake. Seems too good to be true.

**Comment by AlexGalays** [comment]
@gcanti 

`clean` & co have been deprecated but they look useful to me; Is there anywhere I can read about why it's deprecated?

**Comment by gcanti** [comment]
@AlexGalays They were intended to mitigate the exponential growth of the generated declaration files. Recent versions don't suffer from this problem anymore so they have been deprecated.
